api/v1beta1/azuremanagedmachinepool_webhook.go

```diff
@@ -389,3 +397,455 @@ func (m *AzureManagedMachinePool) validateKubeletConfig() error {
+				m.Spec.LinuxOSConfig.SwapFileSizeMB,
+				fmt.Sprintf("SwapFileSizeMB must be at least %d", minSwapFileSizeMB))
+		}
+	}
+
+	if m.Spec.LinuxOSConfig.TransparentHugePageDefrag != nil {
+		switch *m.Spec.LinuxOSConfig.TransparentHugePageDefrag {
+		case TransparentHugePageAlways:
+		case TransparentHugePageDefer:
+		case TransparentHugePageDeferMadvise:
+		case TransparentHugePageMadvise:
+		case TransparentHugePageNever:
+		default:
+			return field.Invalid(
+				field.NewPath("Spec", "LinuxOSConfig", "TransparentHugePageDefrag"),
+				m.Spec.LinuxOSConfig.TransparentHugePageDefrag,
+				fmt.Sprintf("%s, is not a supported configuration", *m.Spec.LinuxOSConfig.TransparentHugePageDefrag))
+		}
+	}
+
+	if m.Spec.LinuxOSConfig.TransparentHugePageEnabled != nil {
+		switch *m.Spec.LinuxOSConfig.TransparentHugePageEnabled {
+		case TransparentHugePageAlways:
+		case TransparentHugePageMadvise:
+		case TransparentHugePageNever:
+		default:
+			return field.Invalid(
+				field.NewPath("Spec", "LinuxOSConfig", "TransparentHugePageEnabled"),
+				m.Spec.LinuxOSConfig.TransparentHugePageEnabled,
+				fmt.Sprintf("%s, is not a supported configuration", *m.Spec.LinuxOSConfig.TransparentHugePageEnabled))
+		}
+	}
+
+	if m.Spec.LinuxOSConfig.Sysctls != nil {
+		if fileSystemSettingsErrors := m.validateLinuxOSConfigFS(); fileSystemSettingsErrors != nil {
+			return fileSystemSettingsErrors
+		}
+
+		if networkCoreSettingsErrors := m.validateLinuxOSConfigNetworkCore(); networkCoreSettingsErrors != nil {
+			return networkCoreSettingsErrors
+		}
+
+		if networkIPv4SettingsErrors := m.validateLinuxOSConfigNetworkIPv4(); networkIPv4SettingsErrors != nil {
+			return networkIPv4SettingsErrors
+		}
+
+		if networkIPv4NeighSettingsErrors := m.validateLinuxOSConfigNetworkIPv4Neigh(); networkIPv4NeighSettingsErrors != nil {
+			return networkIPv4NeighSettingsErrors
+		}
+
+		if networkNetfilterSettingsErrors := m.validateLinuxOSConfigNetworkNetfilter(); networkNetfilterSettingsErrors != nil {
+			return networkNetfilterSettingsErrors
+		}
+
+		if kernelSettingsErrors := m.validateLinuxOSConfigKernel(); kernelSettingsErrors != nil {
+			return kernelSettingsErrors
+		}
+
+		if virtualMemorySettingsErrors := m.validateLinuxOSConfigVM(); virtualMemorySettingsErrors != nil {
+			return virtualMemorySettingsErrors
+		}
+	}
+
+	return nil
+}
+
+// validateLinuxOSConfigFS validates file system settings of Linux OS Config.
+// Refer: https://learn.microsoft.com/en-us/azure/aks/custom-node-configuration#file-handle-limits
+func (m *AzureManagedMachinePool) validateLinuxOSConfigFS() error {
+	if m.Spec.LinuxOSConfig == nil || m.Spec.LinuxOSConfig.Sysctls == nil {
+		return nil
+	}
+
+	if m.Spec.LinuxOSConfig.Sysctls.FsAioMaxNr != nil {
+		fsAioMaxNrLowerBound := int32(65536)
+		fsAioMaxNrUpperBound := int32(6553500)
+		if *m.Spec.LinuxOSConfig.Sysctls.FsAioMaxNr < fsAioMaxNrLowerBound || *m.Spec.LinuxOSConfig.Sysctls.FsAioMaxNr > fsAioMaxNrUpperBound {
```

Comment:
For these checks with upper and lower bounds, would it make the code cleaner to refactor the code into a helper function? I noticed that there are a lot of webhook checks involving upper/lower bounds.

Reply:
Ah that makes sense. If that's the case, it would be safe to remove all the validation from the webhook?

Reply:
Yeah, that makes sense, removed all the validation from the webhook as the functionality is prebaked by the CRD markers.